Fix: Improve code coverage (fixes #2590)

ilyavolodin opened the pull request:
Fixes #2590 This should give us a little bit of breathing room around branch coverage. This is mostly artificial tests for things that shouldn't happen, as well as some `istanbul ignore` comments for things that are either untestable or too hard to test.
In general, we have a bunch of cases where we do branching just for sanity checking. Some of the things that I commented can't possibly happen. I'm not really sure if we should just remove those branches all together.

ilyavolodin commented on lib/config.js:
```diff
@@ -380,6 +385,7 @@ Config.prototype.getConfig = function (filePath) {
         config = util.mergeConfigs(config, this.useSpecificConfig);
     }
 
+    /* istanbul ignore else env is always set to either command line or empty object */
```
This, for example. It's not possible to hit else branch. Ever. Since the code few lines above sets it to either user input or empty object, which will not pass this check.

ilyavolodin commented on lib/formatters/tap.js:
```diff
@@ -66,6 +66,7 @@ module.exports = function(results) {
                 // The first error will be logged as message key
                 // This is to adhere to TAP 13 loosely defined specification of having a message key
                 if ("message" in diagnostics) {
+                    /* istanbul ignore if this is never going to happen we are not setting messages */
```
Or this one. We are creating `diagnostics` object just few lines above, and we are not adding `messages` property to it, so this if statement is impossible to hit.

lo1tuma commented:
1. "too hard to test" is not a valid excuse for not testing the functionality. It is probably a sign that the implementation is bad and should be refactored.
2. If there is some code that can never be reached you should remove this code and NOT just ignoring the coverage of this part.

I think we should enable the `no-warning-comments` rule for eslint and warn for the term `"istanbul ignore"`.

Looking at the coverage report there are a lot of rules with lacking some tests, which would be very easy to fix (e.g. [`eol-last`](https://coveralls.io/builds/2624800/source?filename=lib%2Frules%2Feol-last.js).

ilyavolodin commented:
@lo1tuma In general I agree with "too hard" statement, however, there are legitimate cases for it too. For example testing fallback to user's home directory .eslintrc is actually pretty tough. Testing Windows path processing is pretty much impossible, since we are running tests on Linux. Testing plugin config caching is possible, but is actually pretty complicated (you need to run eslint on multiple files, problem is, when you do, you can't verify that config was actually cached, nothing is exposed at that point) and not worth the effort (cause even if cache isn't going to be used, everything is still going to work).
For point two - that was the question I was asking. I didn't want to just go ahead and remove the code, so I figured I'll comment it out, submit pull request, and see what other's think.

P.S. `eol-last` is a bad example, due to the fact that the only uncovered line is impossible to test. eslint.js will never run on empty files, so uncovered if statement will never be hit (as the istanbul comment says in my pull request).

lo1tuma commented:
As we are talking about unit tests, we should never have a situation where the current system environment has any effect on our tests. I think we already have [a test](https://github.com/eslint/eslint/blob/ae2e8aa7dcf1506c2f5aede531fe14c54df8a588/tests/lib/cli-engine.js#L833) for the plugin cache.
Anyway I will take a more detailed look at the uncovered code of eslint this weekend. I think it shouldn’t be too hard to write some tests for the most cases.

About point 2, I would still prefer to remove the code. If it breaks anything we can revert the comment and provide a test for the code because then we have a proof that it actually is reachable. Same for `eol-last` if this code is not reachable we can safely remove it.

nzakas commented:
I'm generally pretty weary of telling Istanbul to ignore things, so please double check that there's no good way to try, or if some code is unreachable, it's better to remove it.

ilyavolodin commented:
Sounds good. I will remove the code that is unreachable and cherry pick only added tests (without istanbul ignore). I wasn't 100% sure about this pull request myself, and was looking for feedback.

ilyavolodin commented:
I updated pull request and removed all of the `/* istanbul ignore */` flags. I also removed unused code. I was conservative with removal. I left the code that I believe is never going to be hit in `eol-last` because there was a specific issue for this #1007 Although, I think eslint core change since then, and will not process empty files anymore (I remember a heated discussion about it, but I do not remember the outcome of that discussion).

lo1tuma commented:
LGTM
